build.sbt

```diff
@@ -250,17 +261,55 @@ val sparkSettings = Seq(
     val filename = s"$pkgName.tgz"
     val distUrl = url(s"${sparkDistUrl(distVersion)}/$filename")
     val destDir = baseDir / pkgName
-    if (!destDir.exists()) {
+    // It seems that this Tests.Setup block gets run concurrently, which can sometimes cause weirdness to happen.
+    // So we try to use a lockfile to ensure that it only ever runs once
+    // (Yes there still the possibility of a race condition here, but I don't know how to properly synchronize SBT tasks...)
+    val lockFile = baseDir / s"spark_${scalaBinaryVersion.value}_test_setup_is_running.lock"
+    if (lockFile.exists()) {
+      println(s"Lock file $lockFile exists, test setup is already running. Waiting for it to finish...")
+      val start = System.currentTimeMillis()
+      val timeout = 10 * 60 * 1000 // 10 minutes
+      val checkInterval = 10 * 1000 // 10 seconds
+      while (System.currentTimeMillis() < start + timeout && lockFile.exists()) {
+        println(s"Lock file $lockFile still exists after ${System.currentTimeMillis() - start}ms. Waiting...")
+        Thread.sleep(checkInterval)
+      }
+      if (lockFile.exists()) {
+        throw new Exception(s"Lock file $lockFile still exists after $timeout ms. Aborting.")
+      } else {
+        println("Lock file no longer exists, test setup must have finished")
+      }
+    } else {
       baseDir.mkdirs()
-      val pkgFile = baseDir / filename
-      if (!pkgFile.exists()) {
-        pkgFile.createNewFile()
-        println(s"Downloading $distUrl to $pkgFile...")
-        (distUrl #> (baseDir / filename)).!
+      lockFile.createNewFile()
+
+      if (destDir.exists()) {
+        println(s"$destDir already exists, skipping download and extract")
+      } else {
+        val pkgFile = baseDir / filename
+        if (!pkgFile.exists()) {
+          pkgFile.createNewFile()
+          println(s"Downloading $distUrl to $pkgFile...")
+          (distUrl #> pkgFile).!!
+        }
+
+        println(s"Verifying checksum for $pkgFile for $distVersion...")
+        val expectedChecksum = sparkChecksums(distVersion)
+        val actualChecksum = Seq("sha512sum", pkgFile.toString).!!.trim.split(" ").head
+        if (actualChecksum == expectedChecksum) {
+          println(s"Checksum verified for $pkgFile for $distVersion")
+        } else {
+          throw new Exception(s"Checksum mismatch for $pkgFile for $distVersion. Expected:\n$expectedChecksum\nGot:\n$actualChecksum")
+        }
+
+        println(s"Extracting $pkgFile to $baseDir")
+        println(Seq("tar", "-zxpf", pkgFile.toString, "-C", baseDir.toString).!!)
       }
-      println(s"Extracting $pkgFile to $baseDir")
-      Seq("tar", "-zxpf", (baseDir / filename).toString, "-C", baseDir.toString).!
+
+      lockFile.delete()
```

Comment:
I think you also have to set up `lockFile.deleteOnExit()` right after you create it, or you could end up with lockfiles that don't get deleted if there's an exception or what have you.

Also, you should wrap everything between create and delete in a `try` and put the delete in `finally`.